onitu/api/__main__.py

```diff
@@ -30,16 +30,134 @@
 escalator = Escalator(session)
 logger = Logger("REST API")
 
+ONITU_READTHEDOCS = "https://onitu.readthedocs.org/en/latest/api.html"
+HOST = escalator.get(u'api:host') or 'localhost'
+PORT = escalator.get(u'api:port') or 3862
+
+try:
+    SECRET = escalator.get(u'api:secret')
+except:
+    SECRET = str(random.getrandbits(256))
+    escalator.put(u'api:secret', SECRET)
 
-@app.hook('after_request')
-def enable_cors():
-    response.headers['Access-Control-Allow-Origin'] = '*'
-    response.headers['Access-Control-Allow-Methods'] = (
-        'PUT, GET, POST, DELETE, OPTIONS'
-    )
-    response.headers['Access-Control-Allow-Headers'] = (
-        'Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token'
-    )
+
+def enable_cors(fn):
+    def _enable_cors(*args, **kwargs):
+        # set CORS headers
+        response.headers['Access-Control-Allow-Origin'] = '*'
+        response.headers['Access-Control-Allow-Methods'] = (
+            'GET, POST, PUT, OPTIONS')
+        response.headers['Access-Control-Allow-Headers'] = (
+            'Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token')
+        response.headers['Access-Control-Allow-Credentials'] = 'true'
+
+        if bottle.request.method != 'OPTIONS':
+            # actual request; reply with the actual response
+            return fn(*args, **kwargs)
+
+    return _enable_cors
+
+
+def get_identity(user, password):
+    """
+    Returns a user's identity if it exists.
+    """
+
+    try:
+        identity = escalator.get(u'users:{}'.format(user))
+    except:
+        return None
+
+    # We remove the hash from what we give the application.
+    if passwordhash.verify(password, identity.pop('passhash')):
+        return identity
+    return None
+
+
+def get_identity_from_request(request=None):
+    """
+    Retrieves a user's identiy from the request.
+    """
+
+    if request is None:
+        request = bottle.request
+
+    try:
+        identity = jwt.decode(request.cookies.get('identity', ''), SECRET)
+    except jwt.DecodeError:
+        auth = bottle.parse_auth(request.headers.get('Authorization', ''))
+        identity = get_identity(*auth) if auth else None
+
+    return identity
+
+
+class auth(object):
```

Comment:
Because that is what I find the easiest way to build decorators that can be used with arguments like `@auth(role=admin)`.